gobblin-audit/src/main/java/gobblin/audit/values/sink/FsAuditSink.java

```diff
@@ -0,0 +1,123 @@
+
+
+/**
+ *  A Hadoop {@link FileSystem} based {@link AuditSink} that writes audit {@link GenericRecord}s to a file on {@link FileSystem}.
+ *  <ul>
+ *  <li> The {@link FileSystem} {@link URI} can be set using key {@link ConfigurationKeys#FS_URI_KEY}, {@link ConfigurationKeys#LOCAL_FS_URI} is used by default.
+ *  <li> All audit files are written under the base path. The base path can be set using key {@link #FS_SINK_AUDIT_OUTPUT_PATH_KEY}.
+ *  The default path is,
+ *  <pre>
+ *  <code>System.getProperty("user.dir") + "/lumos_value_audit/local_audit";</code>
+ *  </pre>
+ *  <li> It uses <code>auditMetadata</code> to build the audit file name and path.<br>
+ *  <b>The layout on {@link FileSystem} - </b>
+ *  <pre>
+ *  |-- &lt;Database&gt;
+ *    |-- &lt;Table&gt;
+ *       |-- P=&lt;PHASE&gt;.C=&lt;CLUSTER&gt;.E=&lt;EXTRACT_ID&gt;.S=&lt;SNAPSHOT_ID&gt;.D=&lt;DELTA_ID&gt;
+ *          |-- *.avro
+ *  </pre>
+ *  </ul>
+ */
+public class FsAuditSink implements AuditSink {
+
+  private static final String FS_SINK_AUDIT_OUTPUT_PATH_KEY = ValueAuditGenerator.VALUE_AUDIT_CONFIG_KEY_PREFIX + "sink.fs.output.path";
+  private static final String FS_SINK_AUDIT_OUTPUT_DEFAULT_PATH = System.getProperty("user.dir") + "/lumos_value_audit/local_audit";
+  private static final String FILE_NAME_DELIMITTER = "_";
+
+  private final FileSystem fs;
+  private final OutputStream auditFileOutputStream;
+  private final DataFileWriter<GenericRecord> writer;
+  private final Closer closer = Closer.create();
+  private final ValueAuditRuntimeMetadata auditMetadata;
+  @Getter
+  private final Path auditDirPath;
+
+  public FsAuditSink(Config config, ValueAuditRuntimeMetadata auditMetadata) throws IOException {
+
+    this.fs = FileSystem.get(URI.create(ConfigUtils.getString(config, ConfigurationKeys.FS_URI_KEY, ConfigurationKeys.LOCAL_FS_URI)), new Configuration());
+    this.auditDirPath = new Path(ConfigUtils.getString(config, FS_SINK_AUDIT_OUTPUT_PATH_KEY, FS_SINK_AUDIT_OUTPUT_DEFAULT_PATH));
+    this.auditMetadata = auditMetadata;
+    this.auditFileOutputStream = closer.register(fs.create(getAuditFilePath()));
+    DataFileWriter<GenericRecord> dataFileWriter = this.closer.register(new DataFileWriter<GenericRecord>(new GenericDatumWriter<GenericRecord>()));
+    this.writer = this.closer.register(dataFileWriter.create(this.auditMetadata.getTableMetadata().getTableSchema(), this.auditFileOutputStream));
+  }
+
+  /**
+   * Returns the complete path of the audit file. Generate the audit file path with format
+   *
+   *  <pre>
+   *  |-- &lt;Database&gt;
+   *    |-- &lt;Table&gt;
+   *       |-- P=&lt;PHASE&gt;.C=&lt;CLUSTER&gt;.E=&lt;EXTRACT_ID&gt;.S=&lt;SNAPSHOT_ID&gt;.D=&lt;DELTA_ID&gt;
+   *          |-- *.avro
+   *  </pre>
+   *
+   */
+  public Path getAuditFilePath() {
+    StringBuilder auditFileNameBuilder = new StringBuilder();
+    auditFileNameBuilder.append("P=").append(auditMetadata.getPhase()).append(FILE_NAME_DELIMITTER).append("C=")
```

Comment:
Might be worth checking if we will hit file name length limits: https://issues.apache.org/jira/browse/HDFS-6055